src/Parallel/NodeLock.hpp

```diff
@@ -8,66 +8,93 @@
+ * sending node is destroyed to avoid misuse, and throws an error if the lock is
+ * used thereafter. When the lock is deserialized, the nodelock is re-created.
  */
-inline CmiNodeLock create_lock() noexcept { return CmiCreateLock(); }
+class NodeLock {
+ public:
+  NodeLock() noexcept {
+    lock_ = CmiCreateLock();
+  }
 
-/*!
- * \ingroup ParallelGroup
- * \brief Free a converse CmiNodeLock. Using the lock after free is undefined
- * behavior.
- */
-inline void free_lock(const gsl::not_null<CmiNodeLock*> node_lock) noexcept {
+  explicit NodeLock(CkMigrateMessage* /*message*/) noexcept {}
+
+  NodeLock(const NodeLock&) = delete;
+  NodeLock& operator=(const NodeLock&) = delete;
+  NodeLock(NodeLock&& moved_lock) noexcept {
+    moved_lock.destroy();
+    lock_ = CmiCreateLock();
+  }
+
+  NodeLock& operator=(NodeLock&& moved_lock) noexcept {
+    moved_lock.destroy();
+    lock_ = CmiCreateLock();
+    return *this;
+  }
+
+  ~NodeLock() noexcept { destroy(); }
+
+  void lock() noexcept {
+  if (UNLIKELY(destroyed_)) {
+    ERROR("Trying to lock a destroyed lock");
+  }
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wold-style-cast"
-  CmiDestroyLock(*node_lock);
+    CmiLock(lock_);
 #pragma GCC diagnostic pop
 }
 
-/*!
- * \ingroup ParallelGroup
- * \brief Lock a converse CmiNodeLock
- */
-inline void lock(const gsl::not_null<CmiNodeLock*> node_lock) noexcept {
+  bool try_lock() noexcept {
+    if (UNLIKELY(destroyed_)) {
+      ERROR("Trying to try_lock a destroyed lock");
+    }
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wold-style-cast"
-  CmiLock(*node_lock);
+    return CmiTryLock(lock_) == 0;
 #pragma GCC diagnostic pop
-}
-
-/// \cond
-constexpr inline void lock(
-    const gsl::not_null<NoSuchType*> /*unused*/) noexcept {}
-/// \endcond
+  }
 
-/*!
- * \ingroup ParallelGroup
- * \brief Returns true if the lock was successfully acquired and false if the
- * lock is already acquired by another processor.
- */
-inline bool try_lock(const gsl::not_null<CmiNodeLock*> node_lock) noexcept {
+  void unlock() noexcept {
+    if (UNLIKELY(destroyed_)) {
+      ERROR("Trying to unlock a destroyed lock");
+    }
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wold-style-cast"
-  return CmiTryLock(*node_lock) == 0;
+    CmiUnlock(lock_);
 #pragma GCC diagnostic pop
-}
+  }
 
-/*!
- * \ingroup ParallelGroup
- * \brief Unlock a converse CmiNodeLock
- */
-inline void unlock(const gsl::not_null<CmiNodeLock*> node_lock) noexcept {
+  void destroy() noexcept {
+    if(destroyed_) {
+      return;
+    }
+    destroyed_ = true;
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wold-style-cast"
-  CmiUnlock(*node_lock);
+    CmiDestroyLock(lock_);
 #pragma GCC diagnostic pop
-}
+  }
+
+  void pup(PUP::er& p) noexcept {  // NOLINT
+    if (p.isUnpacking()) {
+      lock_ = CmiCreateLock();
+    } else if (p.isPacking()) {
```

Comment:
Could you add a comment `// Note that we don't do anything when calculating the size of the message (p.isSizing())`? People might not be aware that there are more than 2 states (I wasn't when I started using Charm++)

Reply:
Oh, I just thought of something -- when checkpointing, I think I shouldn't actually destroy the lock, because my understanding is that charm will pack the objects, but not unpack them again when it 'continues' after the serialization (it will only unpack when restarting from a stashed checkpoint). I think it makes sense to add the destroy back as a safety measure once we get access to the functions that tell us whether the pup is checkpointing/migrating/other.
Is that consistent with your understanding? If so, I'll just drop the destruction from the pup.

Reply:
That is consistent with my understanding... but Nils should confirm. If this is all correct, a little comment saying that the destroy should be added once we can differentiate between different kinds of serialization might be a helpful reminder.

Reply:
Ah yea that's correct. Good catch! We should make an issue with a checklist of all the things that can be updated once Charm 6.11 is out and we have the new features. Could one of you start that?

Reply:
Opened issue #2417

Reply:
Thanks!